scipy/sparse/_index.py

```diff
@@ -32,74 +26,125 @@ class IndexMixin:
+                else:
+                    res = self._get_sliceXslice(row, col)
             elif col.ndim == 1:
-                return self._get_sliceXarray(row, col)
-            raise IndexError('index results in >2 dimensions')
-        elif row.ndim == 1:
-            if isinstance(col, INT_TYPES):
-                self._raise_on_1d_array_slice()
-                return self._get_arrayXint(row, col)
-            elif isinstance(col, slice):
-                return self._get_arrayXslice(row, col)
-        else:  # row.ndim == 2
+                res = self._get_sliceXarray(row, col)
+            else:
+                raise IndexError('index results in >2 dimensions')
+        else:
             if isinstance(col, INT_TYPES):
-                return self._get_arrayXint(row, col)
+                res = self._get_arrayXint(row, col)
             elif isinstance(col, slice):
-                raise IndexError('index results in >2 dimensions')
-            elif row.shape[1] == 1 and (col.ndim == 1 or col.shape[0] == 1):
-                # special case for outer indexing
-                return self._get_columnXarray(row[:,0], col.ravel())
-
-        # The only remaining case is inner (fancy) indexing
-        row, col = _broadcast_arrays(row, col)
-        if row.shape != col.shape:
-            raise IndexError('number of row and column indices differ')
-        if row.size == 0:
-            return self.__class__(np.atleast_2d(row).shape, dtype=self.dtype)
-        return self._get_arrayXarray(row, col)
+                res = self._get_arrayXslice(row, col)
+            # arrayXarray preprocess
+            elif row.ndim == 2 and row.shape[1] == 1\
+                and (col.ndim == 1 or col.shape[0] == 1):
+                # outer indexing
+                res = self._get_columnXarray(row[:, 0], col.ravel())
+            else:
+                # inner indexing
+                row, col = _broadcast_arrays(row, col)
+                if row.shape != col.shape:
+                    raise IndexError('number of row and column indices differ')
+                if row.size == 0:
+                    res = self.__class__(np.atleast_2d(row).shape, dtype=self.dtype)
+                else:
+                    res = self._get_arrayXarray(row, col)
+
+        # package the result and return
+        if isinstance(self, sparray):
+            # handle formats that support indexing but not 1D (lil for now)
+            if self.format == "lil" and len(new_shape) != 2:
+                return res.tocoo().reshape(new_shape)
+            if res.shape != new_shape:
+                return res.reshape(new_shape)
+        return res
 
     def __setitem__(self, key, x):
-        row, col = self._validate_indices(key)
+        index, _ = self._validate_indices(key)
+
+        # 1D array
+        if len(index) == 1:
+            idx = index[0]
+
+            if issparse(x):
+                x = x.toarray()
+            x = np.asarray(x, dtype=self.dtype)
```

Comment:
put this in an `else` block?

Reply:
We need to check/change the dtype even if `x` was sparse to begin with. So I think we should not have the `else` block, but I think the `asarray` is confusing and inefficient here and maybe why the else is in question. We are only changing the dtype of the array in line 120. So I think we should use `x = x.astype(dtype=self.dtype, copy=False)`